test/test_nn.py

```diff
@@ -11392,9 +11391,159 @@ def __init__(self):
                 self.assertEqual(p.grad.to(devices[0]), pe.grad)
 
 
+instantiate_device_type_tests(TestNNDeviceType, globals())
 
 
-instantiate_device_type_tests(TestNNDeviceType, globals())
+# Set multiple situations to improve convolution test coverage
+class TestConv2dExt(TestCase):
+    _test_case1 = [
+        ['group', [1, 2, 6]],
+        ['bs', [2]],
+        ['input_channels', [1, 16, 30, 32, 33, 64]],
+        ['output_channels', [1, 16, 17]],
+        ['paddings', [[0, 0], [1, 1]]],
+        ['dilations', [[1, 1], [2, 2]]],
+        ['strides', [[1, 1], [2, 2]]],
+        ['kernels', [[1, 1], [2, 2], [3, 3], [5, 5], [7, 7]]],
+        ['output_sizes', [[1, 1]]],
+        ['bias', [True, False]],
+    ]
+
+    _test_case2 = [
+        ['group', [2]],
+        ['bs', [1, 4, 16, 64]],
+        ['input_channels', [3]],
+        ['output_channels', [3]],
+        ['paddings', [[1, 1]]],
+        ['dilations', [[1, 1]]],
+        ['strides', [[1, 1]]],
+        ['kernels', [[3, 3]]],
+        ['output_sizes', [[2, 2]]],
+        ['bias' , [True, False]],
+    ]
+
+    _test_case3 = [
+        ['group', [6]],
+        ['bs', [2]],
+        ['input_channels', [3]],
+        ['output_channels', [3]],
+        ['paddings', [[1, 1]]],
+        ['dilations', [[1, 1]]],
+        ['strides', [[1, 1]]],
+        ['kernels', [[3, 3]]],
+        ['output_sizes', [[1, 1], [2, 2], [3, 3], [5, 5], [6, 6], [7, 7], [16, 16], [17, 17], [32, 32],
+                          [33, 33], [64, 64], [128, 128]]],
+        ['bias' , [True, False]],
+    ]
+
+    def _subtensor(self, tensor, dim, group, g):
+        if tensor is None:
+            return None
+        group_size = int(tensor.size(dim) / group)
+        return tensor.narrow(dim, group_size * g, group_size).contiguous()
+
+    def _thnn_conv(self, input, weight, k, bias, s, p, d):
+        if d[0] > 1 or d[1] > 1:
+            return torch._C._nn.slow_conv_dilated2d(input, weight, k, bias, s, p, d)
+        else:
+            return torch._C._nn.thnn_conv2d(input, weight, k, bias, s, p)
+
+    def _thnn_conv_group(self, input, weight, k, bias, s, p, d, group):
+        if group == 1:
+            return self._thnn_conv(input, weight, k, bias, s, p, d)
+        else:
+            outputs = []
+            for g in range(group):
+                input_g = self._subtensor(input, 1, group, g)
+                weight_g = self._subtensor(weight, 0, group, g)
+                bias_g = self._subtensor(bias, 0, group, g)
+                outputs.append(self._thnn_conv(input_g, weight_g, k, bias_g, s, p, d))
+            return torch.cat(outputs, 1)
+
+    def _generate_cases(self):
+        case_combinations = []
+        for case in [self._test_case1, self._test_case2, self._test_case3]:
+            case_combinations += list(product(*map(lambda x: x[1], case)))
+        return case_combinations
+
+    @skipCUDAIfRocm
+    @dtypesIfCUDA(torch.half, torch.float, torch.double)
```

Comment:
Add the @onlyOnCUDAAndCPU decorator, too (this test shouldn't run on XLA, which is @ailzhang's point, but that's OK for now. In the future it can be generalized. You can find this decorator in torch/testing/_internal/common_device_type.py.